File: php/radix_sort.php
<?php

function countingSortByDigit(&$arr, $exp) {
    $n = count($arr);
    $output = array_fill(0, $n, 0);
    $count = array_fill(0, 10, 0);

    // Count occurrences of digits
    for ($i = 0; $i < $n; $i++) {
        $digit = intdiv($arr[$i], $exp) % 10;
        $count[$digit]++;
    }

    // Cumulative count
    for ($i = 1; $i < 10; $i++) {
        $count[$i] += $count[$i - 1];
    }

    // Build output array
    for ($i = $n - 1; $i >= 0; $i--) {
        $digit = intdiv($arr[$i], $exp) % 10;
        $output[$count[$digit] - 1] = $arr[$i];
        $count[$digit]--;
    }

    // Copy output to original array
    for ($i = 0; $i < $n; $i++) {
        $arr[$i] = $output[$i];
    }
}

function radixSort(&$arr) {
    // Find the maximum number
    $max = max($arr);

    // Do counting sort for every digit
    $exp = 1;
    while (intdiv($max, $exp) > 0) {
        countingSortByDigit($arr, $exp);
        $exp *= 10;
    }
}

// Example usage
$arr = [170, 45, 75, 90, 802, 24, 2, 66];
radixSort($arr);
echo "Sorted array: " . implode(", ", $arr) . "\n";

File: RadixSort.java
/*
 * RadixSort.java
 *
 * Description:
 * Radix Sort is a non-comparison sorting algorithm that sorts numbers 
 * digit by digit starting from the least significant digit (LSD) to 
 * the most significant digit (MSD). It uses Counting Sort as a subroutine.
 *
 * Time Complexity: O(n * k) where k is the number of digits
 * Space Complexity: O(n + k)
 * Stable: Yes
 */

public class RadixSort {
    public static void radixSort(int[] arr) {
        if (arr.length == 0) return;

        // Find the maximum number to know number of digits
        int max = arr[0];
        for (int num : arr) {
            if (num > max) max = num;
        }

        // Apply counting sort for every digit (exp = 1, 10, 100, ...)
        for (int exp = 1; max / exp > 0; exp *= 10) {
            countingSortByDigit(arr, exp);
        }
    }

    private static void countingSortByDigit(int[] arr, int exp) {
        int n = arr.length;
        int[] output = new int[n];
        int[] count = new int[10]; // Digits 0–9

        // Store count of occurrences for each digit
        for (int num : arr) {
            int index = (num / exp) % 10;
            count[index]++;
        }

        // Cumulative count
        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        // Build the output array (traverse backwards for stability)
        for (int i = n - 1; i >= 0; i--) {
            int index = (arr[i] / exp) % 10;
            output[count[index] - 1] = arr[i];
            count[index]--;
        }

        // Copy output to original array
        System.arraycopy(output, 0, arr, 0, n);
    }
}
